### ISSUE TITLE
SQL Server Escape Quotes

### ISSUE THREAD

**Issue by Quantastical** [laravel/framework#24123]
- Laravel Version: 5.6.17
- PHP Version: 7.1.16
- Database Driver & Version: php_sqlsrv_71_ts_x64, php_sqlsrv_71_ts_x86

### Description:

When saving an Eloquent model with a value that contains a quote, the bound parameters are escaping the quote with a slash character. SQL Server uses double single-quotes to escape them properly. Thus, a SQL Syntax error is occurring.

### Steps To Reproduce:

1. Configure a WAMP server with the SQL Server driver.
2. Create a model.
3. Save the model with a property value that contains a single quote character.
4. Notice how the model doesn't save an a SQL syntax error is thrown, e.g.:

QueryException
SQLSTATE[42000]: [Microsoft][ODBC Driver 11 for SQL Server][SQL Server]Incorrect syntax near 's'. (SQL: update [plan_part] set [plan_id] = 48, [is_astm_testing_complete] = 0, [supplier_part_weight] = .0308, [hazard_class_type_id] = 42, [supplier_additional_details] = test's, [modified_by] = 1003 where [plan_part_id] = 55826)
....
in Connection.php (line 664)
....
at Connection->runQueryCallback('update [plan_part] set [plan_id] = ?, [is_astm_testing_complete] = ?, [supplier_part_weight] = ?, [hazard_class_type_id] = ?, [supplier_additional_details] = ?, [modified_by] = ? where [plan_part_id] = ?', array('48', 0, '.0308', '42', 'test\'s', 1003, '55826'), object(Closure))in Connection.php (line 624)

The `'test\'s'` should be escaped as `'test''s'` for SQL Server.

Not sure how to handle this.

**Comment by Quantastical** [comment]
Been researching this further and I'm beginning to think it may have more to do with PDO/SQL Server driver than with Laravel, but I don't really know.

After backtracking through the all of the code in the stack trace from the error, I found that Laravel is using PDO bindValue to bind the values to the statement. That's great, but for some reason it is creating a slash instead of a double quote.

As I understand (could be completely wrong), the driver is responsible for making this transformation, since I see nothing in the Laravel code that would indicate it is handled there. I found a blog article somewhat related: https://blogs.msdn.microsoft.com/brian_swan/2011/05/10/using-pdoquote-with-parameterized-queries/

Don't have a solution yet, but wanted to document my findings as I work to figure out how to resolve this issue.

**Comment by staudenmeir** [comment]
Try the `php_pdo_sqlsrv_*` driver from [here](https://github.com/Microsoft/msphpsql/releases/tag/v5.2.0), it works for me.

**Comment by Quantastical** [comment]
@staudenmeir Which version in particular are you wired up to? This is the first time I've ever had to add extensions (I've always used MySQL in the past), so I don't know if I'm doing something wrong.

I am already using drivers from that source, particularly `php_pdo_sqlsrv_71_ts_x64`, `php_pdo_sqlsrv_71_ts_x86`, `php_sqlsrv_71_ts_x64`, `php_sqlsrv_71_ts_x86`.

I'm running Windows 7 on a 64-bit processor, so I'm guessing I can disable the _x86 ones, but I have to be doing something wrong.

I added the `extension=` for the above four and notice there's a _72 version as well. Do I just have to enable the `php_pdo_sqlsrv` one and not the `php_sqlsrv` one? What's the difference?

I notice there are extensions enabled for php_pdo_mysql and php_pdo_sqllite. Do I have to disable those, too?

I'm just trying to understand the correct process for this.

**Comment by staudenmeir** [comment]
I tested it on PHP 7.2, but 7.1 should work the same way.

Are you running the 64 bit version of PHP? Then try it with `php_pdo_sqlsrv_71_ts_x64` enabled and the other three disabled.

You can leave `php_pdo_mysql` and `php_pdo_sqllite` enabled.

**Comment by Quantastical** [comment]
Thanks for the clarification and assistance in my understanding of this problem. When I disabled everything except the `_x64`, it would say driver not found when I try to login to my app. I guess I'm running a 32-bit version of WAMP, so I need the `_x86` driver instead.

So, I disabled the others and only have `php_pdo_sqlsrv_71_ts_x86` extension enabled. The app lets me log in but when I try to save with a single quote in the value, I'm getting the same syntax error I got before and the stack trace is showing the slash escape character again.

Needless to say, but I think it is becoming obvious that there is a problem with my configuration of WAMP and the drivers (and not with Laravel itself). I suppose it is okay to close this issue out even though I haven't resolved it yet.

I don't know what is going on, but my next steps are to possibly try the 7.2 version of PHP with the appropriate drivers, and then possibly a 64-bit version of WAMP.

It's very odd to me because we're having the same problem on our other webservers as well. I would expect the _x86 driver to work the same.

**Comment by staudenmeir** [comment]
Are there any special settings in `config/database.php`? Can you try a fresh Laravel installation?

Did it work before on your webserver?

**Comment by Quantastical** [comment]
Nothing appears out of the ordinary in `config/database.php` (see below). This is a fresh Laravel installation, as I was trying to migrate everything to the latest version. We were running an old PHP, Laravel, and Yajra DataTables, so I went through and updated everything on my local install.

Unfortunately, I don't know if it worked before on the webservers because we only recently (after the upgrade) reported this bug, so I can't be 100% positive.

My database config is below. The only thing "special" would be a timeout setting that is commented out as I was experiencing some lengthy queries and tried extended the timeout time. We were able to improve query response time, so I commented it back out.

`
return [
    'fetch' => PDO::FETCH_OBJ,
    'default' => env('DB_CONNECTION', 'sqlsrv'),
    'connections' => [
        'sqlite' => [
            'driver' => 'sqlite',
            'database' => env('DB_DATABASE', database_path('database.sqlite')),
            'prefix' => '',
        ],
        'mysql' => [
            'driver' => 'mysql',
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '3306'),
            'database' => env('DB_DATABASE', 'forge'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'charset' => 'utf8',
            'collation' => 'utf8_unicode_ci',
            'prefix' => '',
            'strict' => true,
            'engine' => null,
        ],
        'pgsql' => [
            'driver' => 'pgsql',
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '5432'),
            'database' => env('DB_DATABASE', 'forge'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'charset' => 'utf8',
            'prefix' => '',
            'schema' => 'public',
            'sslmode' => 'prefer',
        ],
        'sqlsrv' => [
            'driver'   => 'sqlsrv',
            'host'     => env('DB_HOST', 'localhost'),
            'database' => env('DB_DATABASE', 'forge'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'charset'  => 'utf8',
            'prefix'   => '',
            'options' => [
              // PDO::SQLSRV_ATTR_QUERY_TIMEOUT => 500
            ],
        ],
        'users' => [
            'driver'   => 'sqlsrv',
            'host'     => env('DB_HOST', 'localhost'),
            'database' => env('DB_DATABASE2', 'forge'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'charset'  => 'utf8',
            'prefix'   => '',
        ],
        'gmpfep' => [
            'driver'   => 'sqlsrv',
            'host'     => env('DB_HOST', 'localhost'),
            'database' => env('DB_DATABASE3', 'forge'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'charset'  => 'utf8',
            'prefix'   => '',
        ],
    ],
    'migrations' => 'migrations',
    'redis' => [
        'cluster' => false,
        'default' => [
            'host' => env('REDIS_HOST', '127.0.0.1'),
            'password' => env('REDIS_PASSWORD', null),
            'port' => env('REDIS_PORT', 6379),
            'database' => 0,
        ],
    ],
];
`

As much as it sucks, I think I'm going to have to try like you say and start from scratch, adding one small piece at a time.

**Comment by Quantastical** [comment]
Well, I wasn't able to solve my problem, but I can effectively rule out Laravel from being the issue.

I created a test site that contains a single PHP file with a very basic query:

```
$pdo = new PDO("sqlsrv:server=$host;Database=$db", $user, $pass);
$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

try {
	$stmt = $pdo->prepare('UPDATE my_table SET notes = ? WHERE id = ?');
	$stmt->execute(["test's",123]);
} catch (Exception $e) {
	var_dump($e);
}
```

It still yields the same error that I was receiving before, "[Microsoft][ODBC Driver 11 for SQL Server][SQL Server]Incorrect syntax near 's'."

I tried switching over to the 64-bit version of WAMP with the appropriate drivers for the version of PHP and still got the error. I tried various configurations of the extensions provided by Microsoft (the php_pdo_* and the _nts, _ts variants) and all of them failed to get recognized (driver not found) except for php_pdo_sqlsrv_71_ts (my WAMP is running PHP 7.1.9.

The php_pdo_..._ts one yielded the same error as before. So, I'm guessing that it's not a Laravel issue, and possibly not even a driver issue since @staudenmeir reports no issues with their implementation.

My next thought might be some sort of database configuration, but that is going well beyond my understanding.

I'm going to close this issue out and move my request over to the StackOverflow Q&A site. If any future visitors stumble upon these notes, I started one question at https://stackoverflow.com/questions/50192310/ and I will try to modify that to report my most recent findings.
